088-rebalance-progress-status.md

```diff
@@ -0,0 +1,392 @@
+  "totalDataToMove": 0,
+  "triggeredSelfHealingTaskId": "",
+  "triggeredTaskReason": "No reason provided (Client: 172.17.0.1, Date: 2024-11-15T19:41:27Z)",
+  "triggeredUserTaskId": "0230d401-6a36-430e-9858-fac8f2edde93"
+}
+```
+
+The formulas used to calculate the field value differ for each applicable `KafkaRebalance` state:
+
+#### State: `ProposalReady`
+
+The rebalance has not started yet so no data surrounding partition movement is available yet, so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+
+#### State: `Rebalancing`
+
+The field will be assigned the "non-verbose" JSON payload of the [/kafkacruisecontrol/state?substates=executor](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) endpoint of the Cruise Control REST API.
+
+#### State: `Stopped`
+
+Once a rebalance has been stopped, it cannot be resumed.
+However, the `executorState.json` information from when before the rebalance was stopped may still be of value to users.
+Therefore, we reuse the same value of `executorState.json` from the previous update.
+
+#### State: `NotReady`
+
+Once a rebalance has been interrupted or completed with errors, it cannot be resumed.
+However, the `executorState.json` information from when before the rebalance was interrupted or completed may still be of value to users.
+Therefore, we reuse the same value of `executorState.json` from the previous update.
+
+#### State: `Ready`
+
+The rebalance is complete in this state so there is no data surrounding ongoing partition movement available, so we omit this field in the `ConfigMap` of the `KafkaRebalance` resource for this state.
+
+### Progress Update Cadence
+
+For ease of implementation and minimizing the load on the CruiseControl REST API server, the operator will only query the `/kafkacruisecontrol/state?substates=executor` endpoint and update the `ConfigMap` upon `KafkaRebalance` resource reconciliation.
+
+In the event that Cruise Control runs into an error when rebalancing, when operator transitions the `KafkaRebalance` resource to the `NotReady` state, it will leave the `progress` section and `ConfigMap` as is.
+In the event that the Cruise Control REST API returns an error or fails to respond to the operator when querying the `/kafkacruisecontrol/state?substates=executor` endpoint during a rebalance, the operator will add a "Warning" condition with error message from the failed REST API call but leave the existing progress section and progress `ConfigMap` as is.
+
+When Cruise Control state retrieval fails, the `KafkaRebalance` resource will be updated as follows:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  ...
+spec: {}
+status:
+  conditions:
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+    status: "True"
+    type: Rebalancing
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
```

Comment:
This might be unlikely case so maybe does not matter, but if Cruise Control REST API returns an error in several reconciliations, would `lastTransitionTime` be updated in  every reconciliation or only if the warning message/state changed? I was wondering how this reflects on the progress field to indicate how out of date `estimatedTimeToCompletionInMinutes` might be.

Reply:
>  would lastTransitionTime be updated in every reconciliation or only if the warning message/state changed

This is a good point. In the event Cruise Control REST API returns an error for several reconciliations, the condition would be updated only when the warning message/state changed. This would highlight  error ongoing errors would better indicate how stale `estimatedTimeToCompletionInMinutes` is.

Let me add a line to clarify this.

Reply:
@tinaselenge Does the added line make this more clear?